docs/source/1.0/features/attention.md

```diff
@@ -1 +1,398 @@
+
+`sink_token_length` is also used to enable this feature.
+
+### Beam-Search
+
+The attention operator supports beam-search. In the context phase, a single
+beam is computed per input sequence. In the generation phase, the MHA/MQA/GQA
+kernel uses an additional tensor to reconstruct the correct path for each beam.
+That tensor is called the `cache_indirection`. Its shape is `[batch_size,
+beam_width, max_seqlen]`.
+
+For a sequence `si`, a beam `bi` and a token `ti`, the element
+`cache_indirection[si][bi][ti]` is an integer between `0` and `beam_width-1`
+that indicates which path in the beam to read the K and V elements from in the
+KV cache. This tensor is populated in the sampling stage.
+
+### Input QKV tensor
+
+The input QKV tensor packs the Q, K and V tensors (concatenated along the last
+dimension) after the projection of the hidden states. It is a 3D tensor. RoPE
+and quantization to INT8 or FP8 (when needed) are performed by the GPT
+attention operator.
+
+In packed mode, its shape is `[num_tokens, 3 * hidden_dim]` where
+`num_tokens` is the total number of tokens in the batch. For the sequences in
+context phase, the number of tokens of a sequence corresponds to its input
+length (even if the beam width is greater than `1` for beam search). For the
+sequences in generation phase, there are `beam_width` tokens per sequence. The
+beam width can be different for each sequence.
+
+In other words, the pseudo-code to compute the number of tokens is:
+
+```python
+num_tokens = 0
+
+# Add the length of each sequence in context phase.
+for seq in context_phase:
+    num_tokens += seq.length
+
+# Add the width of the beam for each sequence in generation phase.
+for seq in generation_phase:
+    num_tokens += seq.beam_width
+```
+
+### Rotary Positional Embedding (RoPE)
+
+The attention operator can perform the computation of the Rotary
+Positional Embedding (RoPE). When that operation is enabled,
+`rotary_embedding_dim` is set to a value greater than 0, it is fused with other
+operations. The GPT operator supports GPT-NeoX and GPT-J forms of RoPE by
+setting `position_embedding_type` to `PositionEmbeddingType.rope_gpt_neox`
+or `PositionEmbeddingType.rope_gptj`.
+
+### ALiBi
+
+The attention operator can apply ALiBi to the result of the `Q*K^T`
+product. The bias is computed on-the-fly from the ALiBi slopes in the optimized
+kernel.
+
+### Scaling factor(s)
+
+In MHA, the output of the `Q*K^T` product is scaled by a constant value that
+is computed as:
+
+```
+norm_factor = 1.f / (q_scaling * sqrt(head_size)).
+```
+
+### Cross Attention
+
+On top of the MHA as self attention needed by GPT-style decoder-only models, the attention operator also supports cross attention.
+
+This enables using the attention operator in a broader aspect as a generic decoder component. For example, the Encoder-Decoder model uses it to issue both the self attention and cross attention modules in its Decoder.
```

Comment:
"This enables the attention operator to be more broadly used as a generic decoder component."